### ISSUE TITLE
docker system prune removes Hyper-V and WSL2 networks

### ISSUE THREAD

**Issue by isanych** [moby/moby#41013]
**Description**
docker system prune removes Hyper-V and WSL2 networks

**Steps to reproduce the issue:**
```
docker system prune --all
...
Deleted Networks:
Default Switch
WSL
```

**Describe the results you received:**
non docker networks are gone, Hyper-V and WSL2 and unusable.

**Describe the results you expected:**
only networks created by docker should be deleted

**Additional information you deem important (e.g. issue happens only occasionally):**
Issue was reported long time ago - https://github.com/docker/for-win/issues/1174 but was ignored

**Output of `docker version`:**

```
Client: Docker Engine - Community
 Version:           19.03.8
 API version:       1.40
 Go version:        go1.12.17
 Git commit:        afacb8b
 Built:             Wed Mar 11 01:23:10 2020
 OS/Arch:           windows/amd64
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          19.03.8
  API version:      1.40 (minimum version 1.24)
  Go version:       go1.12.17
  Git commit:       afacb8b
  Built:            Wed Mar 11 01:37:20 2020
  OS/Arch:          windows/amd64
  Experimental:     true
```

**Output of `docker info`:**

```
Client:
 Debug Mode: false

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 1
 Server Version: 19.03.8
 Storage Driver: windowsfilter (windows) lcow (linux)
  Windows:
  LCOW:
 Logging Driver: json-file
 Plugins:
  Volume: local
  Network: ics internal l2bridge l2tunnel nat null overlay private transparent
  Log: awslogs etwlogs fluentd gcplogs gelf json-file local logentries splunk syslog
 Swarm: inactive
 Default Isolation: hyperv
 Kernel Version: 10.0 19041 (19041.1.amd64fre.vb_release.191206-1406)
 Operating System: Windows 10 Pro Version 2004 (OS Build 19041.264)
 OSType: windows
 Architecture: x86_64
 CPUs: 12
 Total Memory: 31.92GiB
 Name: beta
 ID: KDC4:AXBE:IEZA:QYBJ:RDPR:YGDB:U6NR:7DSX:VQRN:L3IK:L7YI:W5J7
 Docker Root Dir: D:\Docker
 Debug Mode: true
  File Descriptors: -1
  Goroutines: 30
  System Time: 2020-05-22T10:39:07.8021543+01:00
  EventsListeners: 2
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: true
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
 Product License: Community Engine
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Any windows environment

**Comment by thaJeztah** [comment]
@simonferquel @StefanScherer ptal

@isanych I'm not too familiar with the Windows networking stack; are these networks setup by Docker Desktop, or default networks that are setup by Windows?

**Comment by isanych** [comment]
Default Switch created by Hyper-V (and you can create other networks there too), WSL2 created for WSL2 (WSL2 uses Hyper-V). I guess docker uses the same networks as Hyper-V, but it should distinguish networks created by docker somehow and do not delete networks created by other services.

**Comment by thaJeztah** [comment]
Thanks for the extra info!

> I guess docker uses the same networks as Hyper-V, but it should distinguish networks created by docker somehow and do not delete networks created by other services.

Yes, I think that would make sense. Either some metadata should be attached to the network (if possible), or some local storage should be kept by the daemon to know which networks were created by docker. I recall @simonferquel made some changes related to "filtering" / "excluding" some networks (in https://github.com/moby/moby/pull/33053 and https://github.com/moby/moby/pull/37774), but don't think that's directly related.


@isanych do such networks (networks created manually / out of bound) show up when running `docker network ls`, and can they be used by containers?

Trying to describe my concern: if we would be filtering networks based on metadata we add (and/or keeping a local list of networks we "own" / "created"), that would mean we may be causing a breaking change for users that (potentially) rely on that option.

So, in that case, we would still have to "list" those networks to be used, but (perhaps as a default), not _prune_ them, and/or disallow _removing_ them (through `docker network rm`). 🤔

**Comment by thaJeztah** [comment]
@dineshgovindasamy @daschott any suggestions?

**Comment by isanych** [comment]
I don't see Hyper-V networks in docker:
```
docker network ls
NETWORK ID          NAME                DRIVER              SCOPE
d0714f770b9d        nat                 nat                 local
cc3854ed6c7e        none                null                local
```
And I don't see docker networks in Hyper-V too.

**Comment by thaJeztah** [comment]
Thanks! So looks like _some parts_ already filter out those networks, but somehow `docker system prune` finds networks that are not listed 🤔

**Comment by StefanScherer** [comment]
Indeed, I can reproducd the effect after switching to Windows Containers.

```
PS C:\Users\vagrant> docker version
Client: Docker Engine - Community
 Version:           19.03.8
 API version:       1.40
 Go version:        go1.12.17
 Git commit:        afacb8b
 Built:             Wed Mar 11 01:23:10 2020
 OS/Arch:           windows/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.8
  API version:      1.40 (minimum version 1.24)
  Go version:       go1.12.17
  Git commit:       afacb8b
  Built:            Wed Mar 11 01:37:20 2020
  OS/Arch:          windows/amd64
  Experimental:     false
PS C:\Users\vagrant> docker system prune
WARNING! This will remove:
  - all stopped containers
  - all networks not used by at least one container
  - all dangling images
  - all dangling build cache

Are you sure you want to continue? [y/N] y
Deleted Networks:
Default Switch
WSL

Total reclaimed space: 0B
```

I am still able to pull new images from Docker Hub after switching back to WSL 2 backend Linux containers. But I see some network issues in WSL 2.

**Comment by thaJeztah** [comment]
I did some quick looking in the source code to find "where" the filtering happens (both `docker system prune` and `docker network remove` _at a glance_ seem to be hitting the same code-path; yet to find where `docker system prune` / `docker network prune` gets those networks from (if they're not listed in `docker network ls`)

**Comment by raesene** [comment]
little bit of a +1 (I know I should just plus the issue) as this bit me in the backside today and broke the networking in my WSL2 install. 

It'd be great to see it fixed.

**Comment by darkvertex** [comment]
Nasty! I just got quite a scare when I accidentally did `docker network prune` and my WSL2 networking stopped working. It came back after a reboot and WSL worked again.

**Comment by nzbart** [comment]
This bug hit me in Docker for Windows 20.10.13, and it's kinda stressful when you see that the following important Hyper-V networks have been deleted, because it's so unexpected:

* External
* WSL
* Default Switch

**Comment by rendevor** [comment]
It still exists on the Docker 24.0.6 (Docker Desktop 4.24.2)
